### ISSUE TITLE
access to undefined variable with funky macro

### ISSUE THREAD

**Issue by vchuravy** [timholy/Revise.jl#309]
```
git clone https://github.com/climate-machine/CLIMA
julia --project=CLIMA
using Revise
using CLIMA
```

produces lots of errors like
```
┌ Error: evaluation error
│   mod = CLIMA.PlanetParameters
│   ex = :(#= /home/vchuravy/src/CLIMA/src/Utilities/PlanetParameters/src/PlanetParameters.jl:20 =# @exportparameter R_d gas_constant / molmass_dryair "Gas constant dry air (J/kg/K)")
│   exception =
│    UndefRefError: access to undefined reference
│    Stacktrace:
│     [1] getindex at ./array.jl:729 [inlined]
│     [2] lookup_var at /home/vchuravy/.julia/packages/JuliaInterpreter/6LimI/src/interpret.jl:8 [inlined]
│     [3] get_return(::JuliaInterpreter.Frame) at /home/vchuravy/.julia/packages/JuliaInterpreter/6LimI/src/interpret.jl:64
│     [4] #methods_by_execution!#14(::Bool, ::Bool, ::Function, ::Any, ::Revise.CodeTrackingMethodInfo, ::Dict{Module,Array{Expr,1}}, ::JuliaInterpreter.Frame) at /home/vchuravy/.julia/packages/Revise/UtBAC/src/lowered.jl:185
│     [5] #methods_by_execution! at ./none:0 [inlined]
│     [6] #methods_by_execution!#9(::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:define,),Tuple{Bool}}}, ::Function, ::Any, ::Revise.CodeTrackingMethodInfo, ::Dict{Module,Array{Expr,1}}, ::Module, ::Expr) at /home/vchuravy/.julia/packages/Revise/UtBAC/src/lowered.jl:49
│     [7] #methods_by_execution! at ./none:0 [inlined]
│     [8] #eval_with_signatures#59(::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Module, ::Expr) at /home/vchuravy/.julia/packages/Revise/UtBAC/src/Revise.jl:344
│     [9] #eval_with_signatures at /home/vchuravy/.julia/packages/Revise/UtBAC/src/Revise.jl:0 [inlined]
│     [10] #instantiate_sigs!#60(::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::OrderedCollections.OrderedDict{Module,OrderedCollections.OrderedDict{Revise.RelocatableExpr,Union{Nothing, Array{Any,1}}}}) at /home/vchuravy/.julia/packages/Revise/UtBAC/src/Revise.jl:352
│     [11] instantiate_sigs! at /home/vchuravy/.julia/packages/Revise/UtBAC/src/Revise.jl:349 [inlined]
│     [12] queue_includes!(::Revise.PkgData, ::Base.PkgId) at /home/vchuravy/.julia/packages/Revise/UtBAC/src/pkgs.jl:156
│     [13] parse_pkg_files(::Base.PkgId) at /home/vchuravy/.julia/packages/Revise/UtBAC/src/pkgs.jl:130
│     [14] _watch_package(::Base.PkgId) at /home/vchuravy/.julia/packages/Revise/UtBAC/src/pkgs.jl:270
│     [15] (::getfield(Revise, Symbol("##22#23")){Base.PkgId})() at ./task.jl:259
└ @ Revise ~/.julia/packages/Revise/UtBAC/src/lowered.jl:52
```

This particular code used to work on Revise 1.0.4

The macro in question does an `@eval` https://github.com/climate-machine/CLIMA/blob/d4e02e4e7925c8358a1160062fda5ca2b0de9bdb/src/Utilities/ParametersType/src/ParametersType.jl#L45

I will try to make this more minimal.

**Comment by simonbyrne** [comment]
Here is a smaller reproducible test case:
https://github.com/simonbyrne/ParamTest
Note that I've removed the `eval`, so that doesn't appear to be the problem.

```
julia> using Revise, ParamTest
[ Info: Recompiling stale cache file /Users/simon/.julia/compiled/v1.1/ParamTest/ud6IV.ji for ParamTest [22ed0676-7cbb-11e9-268f-bd27814c9c1b]

julia> ParamTest.f()
1
# edit ParamTest.jl to give `f() = 2`

julia> ParamTest.f()
┌ Error: evaluation error
│   mod = ParamTest
│   ex = :(#= /Users/simon/test/ParamTest/src/ParamTest.jl:7 =# @parameter c 1.2 "hmm")
│   exception =
│    UndefRefError: access to undefined reference
│    Stacktrace:
│     [1] getindex at ./array.jl:729 [inlined]
│     [2] lookup_var at /Users/simon/.julia/packages/JuliaInterpreter/6LimI/src/interpret.jl:8 [inlined]
│     [3] get_return(::JuliaInterpreter.Frame) at /Users/simon/.julia/packages/JuliaInterpreter/6LimI/src/interpret.jl:64
│     [4] #methods_by_execution!#14(::Bool, ::Bool, ::Function, ::Any, ::Revise.CodeTrackingMethodInfo, ::Dict{Module,Array{Expr,1}}, ::JuliaInterpreter.Frame) at /Users/simon/.julia/packages/Revise/UtBAC/src/lowered.jl:185
│     [5] #methods_by_execution! at ./none:0 [inlined]
│     [6] #methods_by_execution!#9(::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:define,),Tuple{Bool}}}, ::Function, ::Any, ::Revise.CodeTrackingMethodInfo, ::Dict{Module,Array{Expr,1}}, ::Module, ::Expr) at /Users/simon/.julia/packages/Revise/UtBAC/src/lowered.jl:49
│     [7] #methods_by_execution! at ./none:0 [inlined]
│     [8] #eval_with_signatures#59(::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Module, ::Expr) at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:344
│     [9] #eval_with_signatures at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:0 [inlined]
│     [10] #instantiate_sigs!#60(::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::OrderedCollections.OrderedDict{Module,OrderedCollections.OrderedDict{Revise.RelocatableExpr,Union{Nothing, Array{Any,1}}}}) at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:352
│     [11] instantiate_sigs! at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:349 [inlined]
│     [12] maybe_parse_from_cache!(::Revise.PkgData, ::String) at /Users/simon/.julia/packages/Revise/UtBAC/src/pkgs.jl:224
│     [13] handle_deletions at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:473 [inlined]
│     [14] revise() at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:526
│     [15] revise at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:551 [inlined]
│     [16] run_backend(::REPL.REPLBackend) at /Users/simon/.julia/packages/Revise/UtBAC/src/Revise.jl:941
│     [17] (::getfield(Revise, Symbol("##74#76")){REPL.REPLBackend})() at ./task.jl:259
└ @ Revise ~/.julia/packages/Revise/UtBAC/src/lowered.jl:52
2
```

**Comment by simonbyrne** [comment]
okay, i've simplified it even more. The problem actually appears to be the docstring in the macro: if we remove that, it works fine.

(this explains why `Base.@irrational` doesn't have this problem)

**Comment by timholy** [comment]
Thanks @vchuravy and @simonbyrne. Handling the docsystem is a pain. Fix in #310
